server/datastore/mysql/software_test.go

```diff
@@ -1855,3 +1822,221 @@ func testSoftwareByIDIncludesCVEPublishedDate(t *testing.T, ds *Datastore) {
 		}
 	})
 }
+
+func testAllSoftwareIterator(t *testing.T, ds *Datastore) {
+	host := test.NewHost(t, ds, "host1", "", "host1key", "host1uuid", time.Now())
+
+	software := []fleet.Software{
+		{Name: "foo", Version: "0.0.1", Source: "chrome_extensions"},
+		{Name: "foo", Version: "0.0.3", Source: "chrome_extensions"},
+		{Name: "foo", Version: "v0.0.2", Source: "apps"},
+		{Name: "foo", Version: "0.0.3", Source: "apps"},
+		{Name: "bar", Version: "0.0.3", Source: "deb_packages"},
+	}
+	require.NoError(t, ds.UpdateHostSoftware(context.Background(), host.ID, software))
+	require.NoError(t, ds.LoadHostSoftware(context.Background(), host, false))
+
+	foo_ce_v1 := slices.IndexFunc(host.Software, func(c fleet.Software) bool {
+		return c.Name == "foo" && c.Version == "0.0.1" && c.Source == "chrome_extensions"
+	})
+	foo_app_v2 := slices.IndexFunc(host.Software, func(c fleet.Software) bool { return c.Name == "foo" && c.Version == "v0.0.2" && c.Source == "apps" })
+	bar_v3 := slices.IndexFunc(host.Software, func(c fleet.Software) bool {
+		return c.Name == "bar" && c.Version == "0.0.3" && c.Source == "deb_packages"
+	})
+
+	cpes := []fleet.SoftwareCPE{
+		{SoftwareID: host.Software[foo_ce_v1].ID, CPE: "cpe:foo_ce_v1"},
+		{SoftwareID: host.Software[foo_app_v2].ID, CPE: "cpe:foo_app_v2"},
+		{SoftwareID: host.Software[bar_v3].ID, CPE: "cpe:bar_v3"},
+	}
+	_, err := ds.UpsertSoftwareCPEs(context.Background(), cpes)
+	require.NoError(t, err)
+
+	testCases := []struct {
+		q        fleet.SoftwareIterQueryOptions
+		expected []fleet.Software
+	}{
+		{
+			expected: []fleet.Software{
+				{Name: "foo", Version: "v0.0.2", Source: "apps", GenerateCPE: "cpe:foo_app_v2"},
+				{Name: "foo", Version: "0.0.3", Source: "apps"},
+			},
+			q: fleet.SoftwareIterQueryOptions{IncludedSources: []string{"apps"}},
+		},
+		{
+			expected: []fleet.Software{
+				{Name: "foo", Version: "0.0.1", Source: "chrome_extensions", GenerateCPE: "cpe:foo_ce_v1"},
+				{Name: "foo", Version: "0.0.3", Source: "chrome_extensions"},
+				{Name: "bar", Version: "0.0.3", Source: "deb_packages", GenerateCPE: "cpe:bar_v3"},
+			},
+			q: fleet.SoftwareIterQueryOptions{ExcludedSources: []string{"apps"}},
+		},
+		{
+			expected: []fleet.Software{
+				{Name: "foo", Version: "v0.0.2", Source: "apps", GenerateCPE: "cpe:foo_app_v2"},
+				{Name: "foo", Version: "0.0.3", Source: "apps"},
+			},
+			q: fleet.SoftwareIterQueryOptions{ExcludedSources: []string{"deb_packages"}, IncludedSources: []string{"apps"}},
```

Comment:
Does it make sense to allow setting both included and excluded? (in this case the result will always be the same as setting just the included argument)

No need to change this now, just something I realized now after requesting the checks on the included/excluded values :)

Reply:
Makes sense - I'll do it in this PR since we are already validating the query params.